grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -444,41 +449,92 @@ private static String getResponseBody(List<FieldDescriptor> topLevelFields,
     }
 
     @Nullable
-    private static Function<HttpData, HttpData> generateResponseBodyConverter(TranscodingSpec spec) {
-        @Nullable final String responseBody = spec.responseBody;
+    private static Function<AggregatedHttpResponse, AggregatedHttpResponse> generateResponseConverter(
+            TranscodingSpec spec) {
+        // Ignore the spec if the method is HttpBody. The response body is already in the correct format.
+        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getOutputType())) {
+            return httpResponse -> {
+                final JsonNode jsonNode = extractHttpBody(httpResponse.content());
+
+                // Failed to parse the JSON body, return the original response.
+                if (jsonNode == null) {
+                    return httpResponse;
+                }
+
+                // The data field is base64 encoded.
+                // https://protobuf.dev/programming-guides/proto3/#json
+                final String httpBody = jsonNode.get("data").asText();
+                final byte[] httpBodyBytes = Base64.getDecoder().decode(httpBody);
+
+                final ResponseHeaders newHeaders = httpResponse.headers().withMutations(builder -> {
+                    final JsonNode contentType = jsonNode.get("contentType");
+
+                    if (contentType != null && contentType.isTextual()) {
+                        builder.set(HttpHeaderNames.CONTENT_TYPE, contentType.textValue());
+                    } else {
+                        builder.remove(HttpHeaderNames.CONTENT_TYPE);
+                    }
+                    builder.contentLength(httpBodyBytes.length);
+                });
+
+                return AggregatedHttpResponse.of(newHeaders, HttpData.wrap(httpBodyBytes));
+            };
+        }
+
+        @Nullable
+        final String responseBody = spec.responseBody;
         if (responseBody == null) {
             return null;
-        } else {
-            return httpData -> {
-                try (HttpData data = httpData) {
```

Comment:
Question) Can we revive the old behavior to close the `httpData` in case the data is pooled?